### ISSUE TITLE
STYLE add check for future annotations

### ISSUE THREAD

**Issue by MarcoGorelli** [pandas-dev/pandas#41901]
xref https://github.com/pandas-dev/pandas/pull/41892#issuecomment-857834651

**Comment by aacosta13** [comment]
Hey all, this is my first time helping contribute to an open-source project so forgive any of my ignorance here. I'd be willing to work on this issue, just a couple of questions if that's fine. Based on the referenced comment, should I just add that `import` command to every file (except the ones mentioned)? And I'll be sure to look at the contributions guide before doing so, of course. Thanks.

**Comment by MarcoGorelli** [comment]
Hey - the issue is to add a pre-commit check to make sure that `from __future__ import annotations` is always used, see https://github.com/pymc-devs/pymc-examples/blob/01233b6b32f731c83a11689e14a662dfc574b760/.pre-commit-config.yaml#L22-L30 for an example of how to write a pre-commit check which ensures that a pattern is always present. also see the pre-commit docs

**Comment by ForestMars** [comment]
Since this is needed for 1.3 and is a quick fix: [PR #41941 ]

**Comment by ForestMars** [comment]
@aacosta13 Just noticed your earlier comment you were working on this? Sorry I didn't realize, this was just a super quick one

**Comment by aacosta13** [comment]
@ForestMars no worries, was going to work on it and upload my PR tonight once I got home but it’s cool! And like you said it was an easy fix haha

**Comment by simonjayhawkins** [comment]
> Since this is needed for 1.3 and is a quick fix: [PR #41941 ]

The reason we did the changes for 1.3 is we are going to branch soon and wanted to make sure the new style is on the backport branch to reduce the chances of autobackports failing.

I think that task is done and adding it to the other files is just to avoid new future code additions/modifications using Optional, Union, Type, Tuple, List etc in type annotations.

So just adding `from __future__ import annotations` to the modules that don't yet need them is unlikely to create issues with the autobackports.

This task does not therefore need to be done in time for the rc. (and if the PR was just a few additions and changes to the CI with no behavioral changes... which it should be, could be merged during the rc anyway)

**Comment by ForestMars** [comment]
@simonjayhawkins - There are no behavioral changes, however it does seem that we also need to add that import to 120 files that do no have it (excluding tests, etc.), waiting on Marco to confirm that.

**Comment by aacosta13** [comment]
If the import is needed on all files I’ve actually typed up a script that adds the line to all files from when I originally thought that was the scope of this issue, would just need to verify the files that should be excluded. I could submit the PR if needed.

**Comment by ForestMars** [comment]
@aacosta13 that's a +1 for me. Can you share the script in a gist? 

I was actually going to suggest doing that in a separate PR from the pre-commit hook.

**Comment by aacosta13** [comment]
@ForestMars Sorry for the delayed response, hadn't had time to get to my desk till now. But sure, [here](https://gist.github.com/aacosta13/eb1cb4b22ece354c60f0e7aab90137b0) is the link to the gist of the script.

A few things to note: the script is a bit scrappy, I just typed up something quick as I'm sure it's to be discarded after use. It also ignores any styling placements of the import line, as the only check done is to ensure that it avoids pre-pending the import line above all lines in the case of a .py file being a bash script (leading it to be awkwardly placed between other import lines).

Also, I did not exclude any specific files, and simply added the import to all .py files as I was hoping to get more clarification as to which files would not need it added. I did, however, ensure that all the .py files it is being added to do not already have the import line within them.

I'll be happy to make any corrections if needed.

**Comment by ForestMars** [comment]
@aacosta13 Hmm, this is going to put the import statement in the first empty line (which may be part of the docstring for the file.) I don't think that's what you intended...?

**Comment by MarcoGorelli** [comment]
you can use isort to add imports https://pycqa.github.io/isort/docs/configuration/add_or_remove_imports/

**Comment by aacosta13** [comment]
Nice, that makes things a whole lot easier haha. Then I suppose we just do an `isort` on the list of files that don’t have the import line already which should be fairly easy. Were there any other specific files to look out for to exclude from having the line added?

**Comment by ForestMars** [comment]
yes, you can see the list of exclusions in the code on my PR, which is: 

- __init__.py
- api.py
- setup.py
- asv_bench/
- tests/
- web/

**Comment by aacosta13** [comment]
Alright, here's the updated [gist](https://gist.github.com/aacosta13/eb1cb4b22ece354c60f0e7aab90137b0). Excluded the file types listed and used `isort` to add the import to the rest which does a much cleaner job than my previous script. Let me know if there are any other issues, though.

**Comment by ForestMars** [comment]
@aacosta13 - please see the updated list of file exclusions in my PR #41941. The number of files needing the import statement is reduced from ~120 to 55.

**Comment by MarcoGorelli** [comment]
@aacosta13 could you send a PR to @ForestMars 's branch, so this is all done in the same PR to pandas?

**Comment by ForestMars** [comment]
@aacosta13 - LMK if you'd like me to send the PR adding import to those files, to my branch with the pre-commit hook.  Happy to do that if you've don't have time.

**Comment by ForestMars** [comment]
Just checking in here, as I haven't seen any activity on this. Originally I had the impression this was something that needed to go right in.

**Comment by aacosta13** [comment]
@ForestMars sorry, have been tied up with some personal business over the past few weeks. Sorry about not getting back sooner, I can go ahead and jump back in and fix up the issues on my end today

**Comment by MarcoGorelli** [comment]
Feel free to add the imports in your PR @ForestMars - I think you could also add `files: ^pandas/` in the pre-commit config file
